lib/Dialect/RTL/RTLOps.cpp

```diff
@@ -493,6 +519,62 @@ static void print(OpAsmPrinter &p, RTLModuleOp op) {
                   /*printBlockTerminators=*/true);
 }
 
+/// Lookup the generator for the symbol.  This returns null on
+/// invalid IR.
+Operation *RTLModuleGeneratedOp::getGeneratorKindOp() {
+  auto topLevelModuleOp = (*this)->getParentOfType<ModuleOp>();
+  if (!topLevelModuleOp)
+    return nullptr;
+
+  return topLevelModuleOp.lookupSymbol(generatorKind());
+}
+
+static LogicalResult verifyRTLModuleGeneratedOp(RTLModuleGeneratedOp op) {
+  // Check that this instance is inside a module.
+  auto module = dyn_cast<mlir::ModuleOp>(op->getParentOp());
+  if (!module) {
+    op.emitOpError("should be embedded in a top-level module");
+    return failure();
+  }
+
+  auto referencedKind = op.getGeneratorKindOp();
+  if (referencedKind == nullptr)
+    return op.emitError("Cannot find generator definition '")
+           << op.generatorKind() << "'";
+
+  if (!isa<RTLGeneratorSchemaOp>(referencedKind))
+    return op.emitError("Symbol resolved to '")
+           << referencedKind->getName()
+           << "' which is not a RTLGeneratorSchemaOp";
+
+  auto referencedKindOp = dyn_cast<RTLGeneratorSchemaOp>(referencedKind);
+  SmallVector<StringRef, 8> requiredParams;
+  referencedKindOp.getRequiredParams(requiredParams);
+  auto dict = op->getAttrDictionary();
+  for (auto str : requiredParams)
+    if (!dict.get(str))
+      return op.emitError("Missing attribute '") << str << "'";
+
+  return success();
+}
+
+void RTLGeneratorSchemaOp::getRequiredParams(
+    SmallVectorImpl<StringRef> &results) {
+  for (auto param : requiredAttrs())
+    results.push_back(param.dyn_cast<StringAttr>().getValue());
```

Comment:
This doesn't save any lines, but this can alternatively be expressed with something like:
```c++
auto arrayRef = requiredAttrs().getValue();
results.append(arrayRef.begin(), arrayRef.end());
```

Not sure if this was clearer, but I only recently learned you can do it this way.